go-controller: config and docs cleanups, testcase updates, and no logging to file by default

dcbw opened the pull request:
See the individual commit logs for more info on what each one does...

@shettyg @rajatchopra @pecameron @alinbalutoiu

dcbw commented:
@alinbalutoiu hopefully I addressed your comments about default values in the other PR here.

shettyg commented:
Looks like Alin's commit caused a conflict here. I did not think there will be a conflict.

Also, looks like vagrant is failing with this PR with the bootstrapping failing. Let me see if I see anything obvious.

alinbalutoiu commented on docs/config.md:
````diff
@@ -38,37 +45,61 @@ track connections that are initiated from the pods so that the reverse
+needed to use the OVN southbound database API. Only the the ovn-kubernetes
+master needs to specify the 'server' options.
+```
+address=ssl://1.2.3.4:6641
````
Minor nit, should be 6642 there.

alinbalutoiu commented on etc/ovn_k8s.conf:
```diff
@@ -1,15 +1,35 @@
+server-cacert=path/to/server-ca.crt
+
+[ovnsouth]
+address=ssl://1.2.3.4:6641
```
Minor nit, should be 6642 there.

alinbalutoiu commented:
The only conflict seems to be at the import step:
```
++<<<<<<< f11fe2d5db764a8e2b24d37dbec6ff5e4063c8ab
 +      "runtime"
++=======
+       "reflect"
++>>>>>>> go-controller: make config defaults clearer and fix OVN address bug
```
Both packages should be there.

Only a couple minor nits, other than that looks good to me.

alinbalutoiu commented on go-controller/pkg/config/config.go:
```diff
@@ -16,15 +17,31 @@ import (
 	// Kubernetes holds Kubernetes-related parsed config file parameters and command-line overrides
-	Kubernetes KubernetesConfig
+	Kubernetes = KubernetesConfig{
+		APIServer: "http://localhost:8443",
```
That should be the default value, I think it should be *http**s**://localhost:8443*

dcbw replied:
@shettyg @alinbalutoiu yeah, I mean 8080 because we want the unauthenticated port, since by default we don't have/know the certificates necessary for HTTPS.

shettyg commented on go-controller/pkg/config/config_test.go:
```diff
@@ -161,89 +202,161 @@ var _ = Describe("Config Operations", func() {
+				})
+
+			generateTests("the SSL scheme cannot bootstrap the CA certificate",
+				"bootstapping /foo/bar/baz CA certificate failed",
```
s/bootstapping/bootstrapping

shettyg commented on go-controller/pkg/config/config.go:
```diff
@@ -406,32 +424,22 @@ func InitConfig(ctx *cli.Context, defaults *Defaults) error {
 	}
 
-	OvnNorth.ClientAuth, OvnNorth.ServerAuth, err = getOvnAuth(true, &cliConfig.OvnNorth, &cfg.OvnNorth)
+	OvnNorth, err = buildOvnAuth("nb", &cliConfig.OvnNorth, &cfg.OvnNorth, defaults.OvnNorthAddress)
```
The returned value is not being used anywhere. I am missing something here.

shettyg replied:
I also see that the logging from cliconfig is not being copied into config. So any setting to CLI won't work. For e.g., -logfile is not being honored right now.

dcbw replied:
> The returned value is not being used anywhere. I am missing something here.

@shettyg OvnNorth is a global defined near the top.

dcbw replied:
> I also see that the logging from cliconfig is not being copied into config

@shettyg will fix

pecameron commented on docs/config.md:
```diff
@@ -27,6 +27,13 @@ Copy-Item ".\etc\ovn_k8s.conf" -Destination (New-Item "C:\etc" -Type container -
 
 ## Config values
 
+The config file contains common configuration options shared between the various
```
@dcbw - I have started a man page with this material (PR243 docs/ovn_k8s.conf.5)

pecameron commented on go-controller/README.md:
```diff
@@ -16,42 +16,112 @@ Then find the executables here : go-controller/_output/go/bin/
 'ovnkube' is the single all-in-one executable useful for deploying the ovn cni plugin for a kubernetes deployment
 
-'ovn-k8s-cni-overlay' is the cni executable to be planted under /opt/cni/bin so that it can be invoked for each pod event
+'ovn-k8s-cni-overlay' is the cni executable to be placed in /opt/cni/bin so that it can be invoked for each pod event by kubernetes
```
@dcbw In RHEL/fedora its in /usr/libexec/cni/

dcbw commented:
@shettyg I've addressed the issue with master/node ordering by splitting things up and making sure the master bits which do set-connection and set-ssl run before the node bits which boostrap the cert.  Also addressed the defaults issue you mention and added testcases to ensure it doesn't happen again.

@pecameron addressed the docs issue.

PTAL thanks!

rajatchopra commented:
@dcbw gofmt needed on config.go meanwhile.

shettyg commented:
I think we should set the k8s-apiserver, k8s-token, ovn-remote and ovn-nb even for master.  Right now, we don't really need init-node run on master for it to access other nodes. When we do init-master, we create a OVN logical switch on master and a mgmt port etc. With this change, we can no longer access other nodes from master even though logical switch and mgmt port specific to master node gets created with just init-master.

shettyg commented:
The other issue is that if you run init-master with SSL, it fails now because the bootstrapping command looks not to be run for just init-master.

shettyg commented:
Couple of other things.
1. Looks like we are calling ovn-nbctl set-ssl /etc/openvswitch/ovncontroller-privkey.pem /etc/openvswitch/ovncontroller-cert.pem /etc/openvswitch/ovnnb-ca.cert from init-node code too. That is wrong. This probably works at your end because there is ovn-northd running on every node.

2. We were previously only trying to bootstrap using ovn-nbctl and it would work because it supports --bootstrap-ca-cert option. But ovn-sbctl does not. We are currently dependent on ovn-controller to bootstrap ca-cert for us. I think, it makes sense to skip bootstrapping for ovn-sbctl for the time-being.

Basically, the vagrant is broken right now (even without this patch). With this patch, it still does not work either for tcp or ssl. I think the above are the only things to fix.

alinbalutoiu commented on go-controller/pkg/config/config.go:
```diff
@@ -406,32 +456,30 @@ func InitConfig(ctx *cli.Context, defaults *Defaults) error {
+	overrideFields(&Default, &cfg.Default)
+	overrideFields(&Default, &cliConfig.Default)
+	overrideFields(&Logging, &cfg.Logging)
+	overrideFields(&Logging, &cliConfig.Logging)
```
I don't see where Logging.Level is being honored. In my tests it doesn't matter to which value I change it.
We should follow with a call to logrus.SetLevel for that.
Something like this:
```
func setLogLevel(level int) error {
    switch level {
    case 5:
        logrus.SetLevel(logrus.DebugLevel)
        return nil
    case 4:
        logrus.SetLevel(logrus.InfoLevel)
        return nil
    case 3:
        logrus.SetLevel(logrus.WarnLevel)
        return nil
    case 2:
        logrus.SetLevel(logrus.ErrorLevel)
        return nil
    case 1:
        logrus.SetLevel(logrus.FatalLevel)
        return nil
    case 0:
        logrus.SetLevel(logrus.PanicLevel)
        return nil
    default:
        return fmt.Errorf("setLogLevel: given log level is out of range")
    }
}
```

dcbw replied:
@alinbalutoiu it should get called by the individual app, eg ovnkube.go:

```
func runOvnKube(ctx *cli.Context) error {
        if err := config.InitConfig(ctx, nil); err != nil {
                return err
        }

        // Process log flags
        logrus.SetLevel(logrus.Level(config.Logging.Level))
        logrus.SetOutput(os.Stderr)
```

alinbalutoiu replied:
Ah, didn't check ovnkube, this happened to me when using the CNI plugin.
eg ovn-k8s-cni-plugin.go: https://github.com/openvswitch/ovn-kubernetes/blob/master/go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go#L261-L272

Wouldn't it make more sense to do it right after we read the value? This for example also gets ignored since it's debug log and the level will be set only after the function returns. https://github.com/openvswitch/ovn-kubernetes/blob/master/go-controller/pkg/config/config.go#L438-L443

alinbalutoiu commented on go-controller/pkg/config/config.go:
```diff
@@ -50,14 +67,14 @@ type LoggingConfig struct {
+	// ConfDir specifies the CNI config directory in which to write the overlay CNI config file
 	ConfDir string `gcfg:"conf-dir"`
-	Plugin  string `gcfg:"plugin"`
+	// ConfDir specifies the name of the CNI plugin
```
```Plugin``` instead of ```ConfDir```

alinbalutoiu commented on go-controller/README.md:
```diff
@@ -16,42 +16,112 @@ Then find the executables here : go-controller/_output/go/bin/
+  -loglevel int
+     log verbosity and level: 5=debug, 4=info, 3=warn, 2=error, 1=fatal (default: 4)
+  -logfile string
+     path of a file to direct log output to (default: /var/log/ovn-kubernetes.log)
```
Shouldn't the ```(default: /var/log/ovn-kubernetes.log)``` be removed since by default it doesn't log to a file?

alinbalutoiu commented on go-controller/pkg/config/config.go:
```diff
@@ -406,32 +456,30 @@ func InitConfig(ctx *cli.Context, defaults *Defaults) error {
 		logrus.Infof("Parsed config: %+v", cfg)
-
-		if defaults != nil {
-			fillDefaults(&cfg, defaults)
```
This was one of the issues that I encountered recently. The defaults values from external_ids were fetched only if the config file was existing, otherwise it wouldn't fetch the values from external_ids. Good to see it fixed here.

dcbw replied:
@alinbalutoiu I'd like to do testcases for the external-ids, but we need to bring in something like Kube's FakeExec first.  Will work on that in a follow-up, because I want to unit-test the CA bootstrap functionality too.

dcbw commented:
> I think we should set the k8s-apiserver, k8s-token, ovn-remote even for master. Right now, we don't really need init-node run on master for it to access other nodes. When we do init-master, we create a OVN logical switch on master and a mgmt port etc. With this change, we can no longer access other nodes from master even though logical switch and mgmt port specific to master node gets created with just init-master (ovn-controller needs ovn-remote set for it to program master ovs).

@shettyg agreed, will do.

> The other issue is that if you run init-master with SSL, it fails now because the bootstrapping command looks not to be run for just init-master. You likely need to call config.*.ServerAuth.ConfigureAuth() from master too. I guess, it is okay if it is called twice when someone does 'init-master' and 'init-node' at the same time (I have never run them on the same node though)

But isn't the master the thing that actually *sets* the SSL certs for the database?  I mean, the master is what configures the database connection method, so it should already have the CA cert specified in CLI or config file.  Before the master runs, the database will not have SSL set up, and something has to provide the CA when doing set-ssl and that's currentl the master.  Or am I not quite getting how the DB gets configured?

> Couple of other things.
> 
>     Looks like we are calling ovn-nbctl set-ssl /etc/openvswitch/ovncontroller-privkey.pem /etc/openvswitch/ovncontroller-cert.pem /etc/openvswitch/ovnnb-ca.cert from init-node code too. That is wrong. This probably works at your end because there is ovn-northd running on every node.

Yep, that's wrong if that's happening.

>     We were previously only trying to bootstrap using ovn-nbctl and it would work because it supports --bootstrap-ca-cert option. But ovn-sbctl does not. We are currently dependent on ovn-controller to bootstrap ca-cert for us. I think, it makes sense to skip bootstrapping for ovn-sbctl for the time-being.
> 
> Basically, the vagrant is broken right now (even without this patch). With this patch, it still does not work either for tcp or ssl. I think the above are the only things to fix.

dcbw commented:
@shettyg replying to myself on this, but sbctl does not in fact support --bootstrap-ca-cert or --peer-ca-cert.  Removed those in https://github.com/openvswitch/ovs/pull/220

shettyg commented:
>But isn't the master the thing that actually sets the SSL certs for the database? I mean, the master is what configures the database connection method, so it should already have the CA cert specified in CLI or config file. Before the master runs, the database will not have SSL set up, and something has to provide the CA when doing set-ssl and that's currentl the master. Or am I not quite getting how the DB gets configured?

The master uses the set-ssl method to configure the server certificates. I.e certificates that ovsdb-server backing ovn-nb and ovn-sb should be using to accept new connections. It also sets the IP address to use to accept ssl connections only.  Then the code goes ahead and tries to do a 'init-master' which gets into the client mode to configure the databases via ovn-nbctl. Since the servers only accept ssl connections (or unix domain sockets), and ovn-nbctl is being configured to run using ssl in our code, it needs client certificates to connect (which needs bootstrapping).  We can change ovn-nbctl to use unix domain sockets too - and I agree that people can supply the ca cert to use instead of bootstrapping. But if they don't, then it fails. I think we need to rationalize it, but may we can do it in a future commit instead of overloading many things here

 >Are you sure? I only did this for sbctl because its manpage shows it supports --bootstrap-ca-cert.

yeah, I saw that. There is a bug in ovn-sbctl where it does not accept --bootstrap-ca-cert option. Any fix in OVN itself, means that this code will remain broken for longer.

shettyg commented:
If the current code supports ovn-nbctl to use unix domain socket when running in master, I think, we can agree that being the right method. The last time I looked, it was broken - maybe it is fixed now with your recent commit?

dcbw commented:
@shettyg @alinbalutoiu @rajatchopra updated to address all comments

But can one of you test vagrant?  I cannot bring up vagrant with SSL because it requires a shared directory between all the machines for the cert request signing, and the libvirt vagrant config defaults to rsync, not NFS.  And changing that to NFS leads down a rathole of NFS which I don't quite have time to figure out tonight...

shettyg commented:
TCP works. SSL fails - but that is just due to the fact that we are trying to bootstrap ovn-sbctl - I guess, you expected ovn-controller to take care of it before calling ovnkube (I was expecting that we will skip ovn-sbctl bootstrap and let ovn-controller restart by ovnkube bootstrap it.). But that is okay. I will send an incremental to likely just error message bootstrap error instead of returning error. and let the first actual use of the utility fail.
